### ISSUE TITLE
Mapping generic types breaks inference/narrowing

### ISSUE THREAD

**Issue by jaredru** [Microsoft/TypeScript#16267]
**TypeScript Version:**  2.3.4

**Code**

```ts
interface I {
    x: number | null;
}

function generic<T extends I>(a: T, b: Partial<T>) {
    if (a.x != null) {
        a.x.toFixed(2); // OK
    }

    if (b.x != null) {
        b.x.toFixed(2); // ERROR: Property 'toFixed' does not exist on type 'T["x"]'
    }
}

function notGeneric(a: I, b: Partial<I>) {
    if (a.x != null) {
        a.x.toFixed(2); // OK
    }

    if (b.x != null) {
        b.x.toFixed(2); // OK
    }
}
```

**Expected behavior:**

The mapped generic code should compile without error.

**Actual behavior:**

The mapped generic code does not compile.

For context, the place I'm seeing this issue is React when interacting with `props`.  The React typings for a `Component` are:

```ts
class Component<P, S> {
    props: Readonly<{ children?: ReactNode }> & Readonly<P>;
    // snip
}
```

**Comment by ahejlsberg** [comment]
Duplicate of #14091. Fixed by #15576.

**Comment by jaredru** [comment]
Ah, I searched but didn't find the dupe.  Sorry about that.  Thanks for the quick follow up!
